scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/daemon/ReplaceAfterTerminateScheduler.java

```diff
@@ -15,18 +15,37 @@
 package com.amazonaws.blox.scheduling.scheduler.engine.daemon;
 
 import com.amazonaws.blox.scheduling.scheduler.engine.SchedulingAction;
+import com.amazonaws.blox.scheduling.state.ClusterSnapshot;
+import com.google.common.collect.ImmutableList;
 import java.util.List;
 import java.util.stream.Collectors;
 
 public class ReplaceAfterTerminateScheduler extends DaemonScheduler {
   public static final String ID = "ReplaceAfterTerminate";
 
   public List<SchedulingAction> schedule(DaemonEnvironment env, ClusterSummary summary) {
-    return summary
-        .getInstances()
-        .stream()
-        .filter(i -> env.hasMatchingTask(summary.tasksForInstance(i)))
-        .map(i -> env.startTaskFor(i))
-        .collect(Collectors.toList());
+    List<ClusterSnapshot.Task> tasksToStop =
+        summary
+            .getInstances()
+            .stream()
+            .flatMap(i -> summary.tasksForInstance(i).stream().filter(env::taskToStop))
+            .collect(Collectors.toList());
```

Comment:
Perhaps avoid collecting this to a list here if we're just going to stream it again in the next statement.

Reply:
Sure! And that'll fit in your next comment on using Stream.concat.